docs/dsl-language.md

```diff
@@ -0,0 +1,133 @@
+# Maroon DSL — Goals, Non‑Goals, and Core Spec (Draft)
+
+This doc describes a small, purpose-built language for Maroon. Code in this DSL compiles to Maroon IR (our “assembler”) and runs on the runtime. The goal is simple: you write business logic; the platform guarantees durable compute by determenistic execution on several nodes.
+
+## Why a DSL
+- Deterministic by default: no hidden time/RNG(random number generator)/float surprises, stable iteration order, one standard way to encode data.
+- Fits Maroon’s model: fibers, queues, futures, and timers are native concepts with deterministic scheduling.
+- Safe to replay: re-running the same history yields the same state and outputs.
+- Easy to check: we can warn/error on unbounded loops, impure code in `pure` functions, and risky waits.
+- Smooth upgrades: state has versions and migrations, so rolling upgrades don’t corrupt data.
+
+## Out of Scope (on purpose)
+- No arbitrary threads/syscalls/FFI(foreign function interface).
+- No host-specific behavior (wall-clock reads, RNG, hash-map iteration order, IEEE float edge cases).
+- No DIY persistence: only schema-defined types go to storage; the runtime handles format and migrations.
+- No implicit I/O: external calls must be declared with types, timeouts, and retry policies.
+
+## How Code Runs
+- Unit of work: a lightweight [fiber](./fiber.md).
+- Communication: named FIFO queues (`send`/`recv`).
+- Time: logical monotonic ms via timers (`after(ms)`), not wall-clock.
+
+## External Effects
+- Three kinds: `pure` (no effects), `timer` (logical), `external(service)` (declared capability).
+- External call must declare: request/response types, idempotency key, timeout, retry/backoff; optional compensation.
+- Observability (logs/metrics/traces) does not change behavior/order.
+
+## Numbers and Data
+- Integers: `I64`, `U64` (optionally `I128` later). Overflow behavior is explicit: checked (default), saturating, or wrapping.
+- Decimals: fixed-point `Decimal{scale}` (no floats/NaNs/inf). Rounding mode is explicit and stable.
+- Text/bytes: `String` (UTF‑8) and `Bytes`.
+- Collections: `Vec<T>` (stable order), `Map<K,V>` and `Set<T>` with keys ordered by their byte encoding. No hash maps/sets.
+- Types: `struct`, `enum`, and `type` aliases with explicit field/variant order.
+
+## Canonical Encoding (why ordering is stable)
+- Every value has one byte representation (platform-independent, invertible). We sort map/set keys by these bytes.
+- Examples: big-endian integers; `Decimal{2}(1)` and `1.00` encode the same; strings use UTF‑8 NFC normalization; no “-0”.
+
+## Fiber State and Persistence
+- No global app state. Each fiber owns its own persistent state, defined inside that fiber.
```

Comment:
(I would add that in the current pre-CRDT pro-total-order world, technically speaking, for each era the "global state" still very much exists; it's just not directly observable in O(1) by any fiber whatsoever.)